### ISSUE TITLE
Inaccurate error message when type has incorrect number of type arguments

### ISSUE THREAD

**Issue by svick** [dotnet/roslyn#24406]
**Version Used**: master (22 Jan 2018; 03839fa)

**Steps to Reproduce**:

Try to [compile the following code](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAN0QSwGYE8A+AAgEwAEAsrgKIAeADgnFFJgPYB2pA3gL4CwAKBIVq9RszbsAPABUAfN34DBwgMKCug0tpG0GTFhymZ2MJKRMwFcANyCeQA===):

```c#
class MyExpression {}
class MyExpression<T> {}

class C
{
    MyExpression<int, int> e;
}
```

**Expected Behavior**: An error message that explains that `MyExpression` expects 0 or 1 type arguments, but was used with 2.

**Actual Behavior**:

> CS0308: The non-generic type 'MyExpression' cannot be used with type arguments

This error message is not accurate: `MyExpression` can be used with type arguments (specifically, with one), it just can't be used with two.

The actual code where I encountered this was using `System.Linq.Expressions.Expression<T>`: the source code contained `Expression<Data, bool>` instead of `Expression<Func<Data, bool>>`. If the error message was clearer, I think I would have realized what the issue was sooner.
